wire-library/wire-runtime-swift/src/main/swift/ProtoReader.swift

```diff
@@ -745,7 +778,66 @@ public final class ProtoReader {
             throw ProtoDecoder.Error.mapEntryWithoutKey(value: value)
         }
         guard let unwrappedValue = value else {
-            throw ProtoDecoder.Error.mapEntryWithoutValue(key: key)
+            throw ProtoDecoder.Error.mapEntryWithoutValue(key: unwrappedKey)
+        }
+
+        // Preallocate a few empty slots to avoid reallocations.
+        if dictionary.isEmpty && MemoryLayout<V>.size <= ProtoReader.collectionPreallocationMaxValueSize {
+            dictionary.reserveCapacity(ProtoReader.collectionPreallocationSlotCount)
+        }
+
+        dictionary[unwrappedKey] = unwrappedValue
+    }
+
+    /// Decode a key-value pair where the value is RawReprentable with RawValue == Int32.
+    /// - Parameters:
+    ///   - dictionary: The dictionary in which to add key-value pair if the decoded value is a known case in V.
+    ///   - decodeKey: A closure that decodes the key for each pair
+    ///   - addUnknownPair: A closure that adds to the key-value pair to unknown fields in the event a given raw value is not a known case in V.
+    private func decode<K, V: RawRepresentable>(
+        into dictionary: inout [K: V],
+        decodeKey: () throws -> K,
+        addUnknownPair: (UInt32, K, V.RawValue) throws -> ()
+    ) throws where V.RawValue == UInt32 {
+        var key: K?
+        var value: V?
+        var rawValue: UInt32?
+
+        guard let startTag = currentTag else {
+            fatalError("Decoding enum map but current tag is not set.")
+        }
+
+        let token = try beginMessage()
+        while let tag = try nextTag(token: token) {
+            switch tag {
+            case 1: key = try decodeKey()
+            case 2:
+                let intValue = try readVarint32()
+                let enumValue = V(rawValue: intValue)
+                if enumValue == nil {
+                    if case .throwError = enumDecodingStrategy {
```

Comment:
nit: I *think* you can just use `if enumDecodingStrategy == .throwError` here, but maybe it needs to conform to `Equatable` for that to work. This is fine too.